pxd.c

```diff
@@ -1520,24 +1592,38 @@ ssize_t pxd_remove(struct fuse_conn *fc, struct pxd_remove_out *remove)
 	pr_info("removing device %llu", pxd_dev->dev_id);
 
 	/* Make sure the req_fn isn't called anymore even if the device hangs around */
-	if (pxd_dev->disk && pxd_dev->disk->queue){
+	if (pxd_dev->exported && pxd_dev->disk && pxd_dev->disk->queue){
 #ifndef __PX_BLKMQ__
 		mutex_lock(&pxd_dev->disk->queue->sysfs_lock);
 
 		QUEUE_FLAG_SET(QUEUE_FLAG_DYING, pxd_dev->disk->queue);
 
 		mutex_unlock(&pxd_dev->disk->queue->sysfs_lock);
+#else
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,13,0)
+		// Do not mark queue dead, del_gendisk will try to
+		// submit all outstanding IOs on this device
 #else
 		blk_set_queue_dying(pxd_dev->disk->queue);
+#endif
 #endif
 	}
 
+	cleanup = pxd_dev->exported;
+	pxd_dev->exported = false;
 	spin_unlock(&pxd_dev->lock);
 
-	pxd_fastpath_reset_device(pxd_dev);
-	device_unregister(&pxd_dev->dev);
+	if (cleanup) {
```

Comment:
Why is the exported check needed? It's not going to protect against concurrent pxd_remove() on the same device.

Reply:
If pxd_remove is run unconditionally without looking at exported flag, there is an assumption that from userspace pxd_remove gets called only if device add succeeds and is exported to the kernel.

If for some reason, only pxd_add got called and no followup pxd_export was called, but pxd_remove now follows from userspace, then it will likely crash the kernel. This though currently does not happen.
So having the check doesnt hurt.